### ISSUE TITLE
redis-rb 3.0 support

### ISSUE THREAD

**Issue by catwell** [soveran/ohm#60]
The title says it all ;)

**Comment by TiagoCardoso1983** [comment]
Is this really fixed? I'm getting the following error:

```ruby
Ohm.redis = ::Redis.connect(url: redis_url)
```

```
[1] pry(main)> Run.all.first
Redis::CommandError: ERR unknown command 'call'
from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/redis-3.2.1/lib/redis/client.rb:113:in `call'
```

Or was this ```#call``` method removed at some point?

**Comment by soveran** [comment]
Hello @TiagoCardoso1983, starting on version 2.0.0, Ohm moved to a more lightweight Redis client called [Redic](https://rubygems.org/gems/redic). You can check how to customize the connection in the [Connecting to a Redis database](https://github.com/soveran/ohm/#connecting-to-a-redis-database) section of the README. Note that you don't need to require any other library, as Ohm takes care of that part.

**Comment by TiagoCardoso1983** [comment]
@soveran I saw it and (for now) have it setup that away, however, as I'm using thin/EM, I'd like something that would be compatible with the reactor. redis-rb already provides this. I haven't seen the full code, but I'd say it would be easier to make ohm compatible with the redis raw client than build eventmachine support in redic. the way I see it, Redic connections respond to a ```call```method. Wouldn't it be easy to "translate" it in the redis-rb connection?

**Comment by soveran** [comment]
Maybe what we can do is to create a library that wraps redis-rb in order to make it API compatible with Redic. I can take a look at that tomorrow morning, and I'll let you know how it goes. Sounds good?

**Comment by TiagoCardoso1983** [comment]
Very :+1 . Thx for the awesome feedback!

**Comment by soveran** [comment]
I tried this and it looks like it works with Ohm, but I haven't tried it with EM, can you check it and let me know?

```ruby
require "redis"
require "ohm"

class RedisWrapper
  URL = "redis://:%{password}@%{host}:%{port}"

  def initialize(redis)
    @redis = redis
    @redic = Redic.new(url)
  end

  def call(*args)
    @redis.synchronize do |client|
      client.call(args)
    end
  end

  def queue(*args)
    @redic.queue(*args)
  end

  def commit
    @redis.synchronize do |client|
      @redic.buffer.each do |args|
        client.write(args)
      end

      @redic.buffer.map do
        client.read
      end
    end
  ensure
    @redic.reset
  end

  def url
    URL % @redis.client.options
  end

  alias call! call
end
```

And here's how to use it:

```ruby
Ohm.redis = RedisWrapper.new(Redis.new(...))
```

**Comment by TiagoCardoso1983** [comment]
humm, wont you have then this way two simultaneous connections open to redis?

**Comment by soveran** [comment]
No, it only uses Redic for its thread-safe buffering features, so it never connects. In both clients the connection is lazy, meaning that it won't be established unless you send a command. This wrapper only sends commands to redis-rb, never to Redic.

**Comment by TiagoCardoso1983** [comment]
I have a few concerns. First would be that envvar being the same url you pass as argument to Redis.new (maybe you did it only for testing, if so, nevermind my comment). The second is the thread-safe buffering features. Is there something I am missing, or doesn't redis-rb already synchronize access to its client across features, thereby already protecting access to the client buffer? Just saying, as the use-case is eventmachine, thread-safe buffering would not be the main concern, Also, the ```commit``` method of yours uses an ```args``` variable which isn't defined everywhere. 

Bottom line, I'm not a big fan of mixing redis-rb and redic, unless the thread-safe buffering features are really really needed. Otherwise I'd just try to reimplement the ```queue``` and ```commit``` using the redis client. I can see the case where redic should be the default for multithreaded scenario, but I think the API should be client-independent. I'll try to come up tomorrow with some sort of workable mixin for my use-case. Nevertheless, thanks for the help.

**Comment by soveran** [comment]
> First would be that envvar being the same url you pass as argument to Redis.new (maybe you did it only for testing, if so, nevermind my comment).

Yes, it's not necesary at all actually it doesn't change anything, because it's not used for any connection. I modified the code so that `Redic.new` doesn't receive the URL. The `url` method is necesary because `Ohm` keeps a cache of which servers have the Lua scripts loaded, and the server URLs are used as cache keys. That's why the wrapper needs this `url` method.

> The second is the thread-safe buffering features. Is there something I am missing, or doesn't redis-rb already synchronize access to its client across features, thereby already protecting access to the client buffer?

The thread-safe buffering is not related to how the we write and read the commands from Redis, but only with how we build the list of commands. You can check the implementation in Redic, just to get an idea of what we are talking about.

> Just saying, as the use-case is eventmachine, thread-safe buffering would not be the main concern, 

Thread safe buffering is a concern even if you use EM, because it has nothing to do with how it interact with sockets.

> Also, the commit method of yours uses an args variable which isn't defined everywhere.

The `commit` method uses an `args` variable that is retrieved from the buffer with an iterator.

> Bottom line, I'm not a big fan of mixing redis-rb and redic, unless the thread-safe buffering features are really really needed. Otherwise I'd just try to reimplement the queue and commit using the redis client.

It's really really needed :-)

But of course you can implement those commands too (please check my comment at the end). My advice would be to read [Redic's code](https://github.com/amakawa/redic/blob/master/lib/redic.rb) to see how it works, because that way you will understand the buffering issue and why it's trivial to reuse that code given the fact that Redic is already loaded when you are using Ohm.

> I can see the case where redic should be the default for multithreaded scenario, but I think the API should be client-independent. I'll try to come up tomorrow with some sort of workable mixin for my use-case. Nevertheless, thanks for the help.

Redic is the default because in my opinion it's a better client. I'm co-author of both Redic and redis-rb, so I'm not basing my preference in some other factor. Ohm doesn't rely on any internal design of the client, it is client agnostic, but the client's API should be similar to that of Redic, which is very simple. 

If you could test the code I sent you and give me feedback, I can consider releasing a gem with a reimplementation of the thread-safe buffer, so in the end it won't use Redic internally.

**Comment by TiagoCardoso1983** [comment]
Thx for the reply, will do :) I checked the source code for both gems, and I do agree that redic is a much lighter fit for any kind of scenario, redis-rb bloats it a bit by redefining every command. Now, if it would only have an EM/Celluloid adapter... :) 

I'll come back with feedback.

**Comment by TiagoCardoso1983** [comment]
first try in the console:

```
[3] pry(main)> EM.run { puts Run.all.first ; EM.stop }
FiberError: can't yield from root fiber
from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/redis-3.2.1/lib/redis/connection/synchrony.rb:81:in `yield'
```

Using it in the thin process is no better:

```
14:20:05 web.1  | /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/redis-3.2.1/lib/redis/connection/synchrony.rb:78:in `resume': double resume (FiberError)
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/redis-3.2.1/lib/redis/connection/synchrony.rb:78:in `block in connect'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/eventmachine-1.0.8/lib/em/deferrable.rb:151:in `call'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/eventmachine-1.0.8/lib/em/deferrable.rb:151:in `set_deferred_status'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/eventmachine-1.0.8/lib/em/deferrable.rb:191:in `succeed'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/redis-3.2.1/lib/redis/connection/synchrony.rb:20:in `connection_completed'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/eventmachine-1.0.8/lib/eventmachine.rb:193:in `run_machine'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/eventmachine-1.0.8/lib/eventmachine.rb:193:in `run'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/thin-1.6.3/lib/thin/backends/base.rb:73:in `start'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/thin-1.6.3/lib/thin/server.rb:162:in `start'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/rack-1.6.4/lib/rack/handler/thin.rb:19:in `run'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/rack-1.6.4/lib/rack/server.rb:286:in `start'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/rack-1.6.4/lib/rack/server.rb:147:in `start'
14:20:05 web.1  |       from /home/taacati1/emskripten/.bundle/ruby/2.0.0/gems/rack-1.6.4/bin/rackup:4:in `<top (required)>'
```
